src/collision/collide_and_slide.rs

```diff
@@ -0,0 +1,389 @@
+                break;
+            };
+            let safe_distance = Self::pull_back(hit, vel_dir, config.skin_width);
+            if !on_hit(CollideAndSlideHitData {
+                hit,
+                position,
+                velocity,
+                safe_distance,
+            }) {
+                velocity = Vector::ZERO;
+                break 'outer;
+            }
+            if hit.distance == 0.0 {
+                // entity is completely trapped in another solid
+                velocity = Vector::ZERO;
+                break 'outer;
+            }
+            time_left -= time_left * (safe_distance / speed);
+
+            position += vel_dir * safe_distance;
+
+            // if this is the same plane we hit before, nudge velocity
+            // out along it, which fixes some epsilon issues with
+            // non-axial planes
+            let mut i = 0;
+            while i < planes.len() {
+                if hit.normal1.dot(planes[i].into()) > (1.0 - DOT_EPSILON) {
+                    velocity += hit.normal1 * config.duplicate_plane_nudge;
+                    break;
+                }
+                i += 1;
+            }
+            if i < planes.len() {
+                continue;
+            }
+            if planes.len() >= config.max_planes {
+                velocity = Vector::ZERO;
+                break 'outer;
+            }
+            planes.push(Dir::new_unchecked(hit.normal1));
+
+            // modify velocity so it parallels all of the clip planes
+
+            // find a plane that it enters
+            for i in 0..planes.len() {
+                let into = velocity.dot(planes[i].into());
+                if into >= 0.0 {
+                    // move doesn't interact with the plane
+                    continue;
+                }
+
+                // slide along the plane
+                #[cfg_attr(feature = "2d", expect(unused_mut, reason = "only used in 3D branch"))]
+                let mut current_clip_velocity = Self::clip_velocity(velocity, &[planes[i]]);
+
+                // see if there is a second plane that the new move enters
+                #[cfg_attr(
+                    feature = "2d",
+                    expect(
+                        clippy::needless_range_loop,
+                        reason = "This variant is cleaner in this case"
+                    )
+                )]
+                for j in 0..planes.len() {
+                    if j == i {
+                        continue;
+                    }
+                    if current_clip_velocity.dot(planes[j].into()) >= DOT_EPSILON {
+                        // move doesn't interact with the plane
+                        continue;
+                    }
+                    #[cfg(feature = "2d")]
+                    {
+                        // stop dead at a double plane interaction
+                        velocity = Vector::ZERO;
+                        break 'outer;
+                    }
```

Comment:
Does this cause hitches on obtuse-angle seams?

Reply:
Ah, I'd been imagining this as top-down 2D, running along hugging a wall. In that case I'd expect this to cause hitching whenever the wall bends inwards: stopping completely where I'd expect (as a player) to just slow down slightly, by an amount depending on how much the wall is bending.

On the other hand, I couldn't figure out how to make the idea that would have changed this work, so I can't be of much use 😅